File: 2015/ruby/day16.rb
def assert_equal(expected, actual, str="")
  raise "Expected #{expected} but got #{actual}. #{str}" unless expected == actual
end

data = File.read('day16.txt')

ticker_tape = <<~TXT
children: 3
cats: 7
samoyeds: 2
pomeranians: 3
akitas: 0
vizslas: 0
goldfish: 5
trees: 3
cars: 2
perfumes: 1
TXT

def parse_line(line)
  m = /^Sue (?<sue>\d+): (?<rest>.*)/.match(line)
  raise "Could not parse #{line}" unless m
  vals = Hash[
    m['rest'].scan(/(?<name>[a-z]+): (?<cnt>\d+)/).map {|(n, c)| [n.to_sym, c.to_i]}
  ]
  {sue: m['sue'].to_i}.merge(vals)
end
def parse_clues(data)
  data.lines.map do |l|
    clue, cnt = l.split(': ')
    [clue.to_sym, cnt.to_i]
  end
end

def find_sue(data, ticker_tape)
  clues = parse_clues(ticker_tape)
  sues = data.lines.map {|l| parse_line(l)}
  sues.find do |sue|
    clues.all? do |(clue, cnt)|
      sc = sue[clue]
      next true if sc.nil?

      case clue
      when :cats, :trees
        sc > cnt
      when :pomeranians, :goldfish
        sc < cnt
      else
        sue[clue] == cnt
      end
    end
  end
end
assert_equal({sue: 1, goldfish: 9, cars:0, samoyeds: 9}, parse_line('Sue 1: goldfish: 9, cars: 0, samoyeds: 9'))
puts find_sue(data, ticker_tape)

File: 2015/go/day16/day16.go
package day16

import (
	"fmt"
	"regexp"
	"strconv"
)

type Sue struct {
	Idx    int
	Values map[string]int
}

func ParseLine(line string) Sue {
	var idx int

	fmt.Sscanf(line, "Sue %d: %s", &idx)
	prefix := fmt.Sprintf("Sue %d: ", idx)
	prefixLength := len(prefix)
	rest := line[prefixLength:]
	r := regexp.MustCompile(`([a-z]+): (\d+)`)
	matches := r.FindAllStringSubmatch(rest, -1)
	sue := Sue{idx, map[string]int{}}
	for _, m := range matches {
		sue.Values[m[1]], _ = strconv.Atoi(m[2])
	}

	return sue
}

func Matches(sue Sue, tickerTape map[string]int, useRange bool) bool {
	for k, v := range tickerTape {
		sv, ok := sue.Values[k]
		if !ok {
			continue
		}
		if v != sv {
			if useRange {
				if k == "cats" || k == "trees" {
					if sv <= v {
						return false
					}
				} else if k == "pomeranians" || k == "goldfish" {
					if sv >= v {
						return false
					}
				} else {
					return false
				}
			} else {
				return false
			}
		}
	}
	return true
}

func FindSue(input []string, tickerTape map[string]int, useRange bool) int {
	for _, line := range input {
		sue := ParseLine(line)
		if Matches(sue, tickerTape, useRange) {
			return sue.Idx
		}
	}
	return -1
}
